File: scripts/fetch-cards.mjs
import { mkdir, readFile, writeFile } from 'node:fs/promises'
import { join } from 'node:path'

const DATA_URL =
  'https://raw.githubusercontent.com/SkAZi/berserk-nxt/main/resources/data.json'
const CONST_URL =
  'https://raw.githubusercontent.com/SkAZi/berserk-nxt/main/resources/const.json'

const OUT_DIR = join(process.cwd(), 'public', 'data')
const IMAGE_MAP_PATH = join(OUT_DIR, 'image-map.json')

function normalizeCard(raw) {
  const alts = raw.alts ?? []
  const hasNonStandardAlt = alts.some((alt) =>
    ['pf', 'altpf', 'fpf', 'altfpf'].includes(alt),
  )

  return {
    id: String(raw.id),
    setId: raw.set_id,
    number: raw.number,
    name: raw.name,
    cost: raw.cost,
    life: raw.life ?? undefined,
    move: raw.move ?? undefined,
    hit: raw.hit?.length ? raw.hit : undefined,
    text: raw.text || undefined,
    type: raw.type,
    color: raw.color,
    rarity: raw.rarity,
    elite: Boolean(raw.elite),
    hasAlt: alts.length > 0,
    nonStandardLayout: hasNonStandardAlt,
  }
}

async function loadImageMap() {
  try {
    const raw = await readFile(IMAGE_MAP_PATH, 'utf-8')
    return JSON.parse(raw)
  } catch {
    return {}
  }
}

async function main() {
  const [dataRes, constRes, imageMap] = await Promise.all([
    fetch(DATA_URL),
    fetch(CONST_URL),
    loadImageMap(),
  ])

  if (!dataRes.ok || !constRes.ok) {
    throw new Error('Failed to download card data')
  }

  const rawCards = await dataRes.json()
  const constData = await constRes.json()

  const cards = rawCards
    .filter((card) => !card.ban)
    .map(normalizeCard)
    .filter((card) => constData.sets[String(card.setId)])
    .map((card) => {
      const imageUrl = imageMap[`${card.setId}/${card.number}`]
      return imageUrl ? { ...card, imageUrl } : card
    })

  await mkdir(OUT_DIR, { recursive: true })
  await writeFile(
    join(OUT_DIR, 'cards.json'),
    JSON.stringify(cards, null, 2),
    'utf-8',
  )
  await writeFile(
    join(OUT_DIR, 'const.json'),
    JSON.stringify(constData, null, 2),
    'utf-8',
  )

  console.log(`Saved ${cards.length} cards to public/data/cards.json`)
}

main().catch((error) => {
  console.error(error)
  process.exit(1)
})

File: scripts/fetch-cards.py
#!/usr/bin/env python3
"""Bootstrap data when Node.js is unavailable."""

import json
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
OUT = ROOT / "public" / "data"
DATA_URL = "https://raw.githubusercontent.com/SkAZi/berserk-nxt/main/resources/data.json"
CONST_URL = "https://raw.githubusercontent.com/SkAZi/berserk-nxt/main/resources/const.json"


def fetch_json(url: str):
    with urllib.request.urlopen(url) as response:
        return json.loads(response.read().decode("utf-8"))


def normalize_card(raw: dict, known_sets: set[str]) -> dict | None:
    if raw.get("ban"):
        return None

    set_id = raw["set_id"]
    if str(set_id) not in known_sets:
        return None

    alts = raw.get("alts") or []
    return {
        "id": str(raw["id"]),
        "setId": set_id,
        "number": raw["number"],
        "name": raw["name"],
        "cost": raw["cost"],
        "life": raw.get("life"),
        "move": raw.get("move"),
        "hit": raw.get("hit") or None,
        "text": raw.get("text") or None,
        "type": raw["type"],
        "color": raw["color"],
        "rarity": raw["rarity"],
        "elite": bool(raw.get("elite")),
        "hasAlt": len(alts) > 0,
        "nonStandardLayout": any(alt in {"pf", "altpf", "fpf", "altfpf"} for alt in alts),
    }


def main() -> None:
    raw_cards = fetch_json(DATA_URL)
    const_data = fetch_json(CONST_URL)
    known_sets = set(const_data["sets"].keys())

    OUT.mkdir(parents=True, exist_ok=True)
    image_map_path = OUT / "image-map.json"
    image_map = {}
    if image_map_path.exists():
        image_map = json.loads(image_map_path.read_text(encoding="utf-8"))

    cards = []
    for raw in raw_cards:
        card = normalize_card(raw, known_sets)
        if card is not None:
            key = f"{card['setId']}/{card['number']}"
            if key in image_map:
                card["imageUrl"] = image_map[key]
            cards.append(card)

    (OUT / "cards.json").write_text(
        json.dumps(cards, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )
    (OUT / "const.json").write_text(
        json.dumps(const_data, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )
    print(f"Saved {len(cards)} cards to {OUT / 'cards.json'}")


if __name__ == "__main__":
    main()
